File: samza-core/src/main/scala/org/apache/samza/metrics/reporter/MetricsSnapshotReporterFactory.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.samza.metrics.reporter

import org.apache.samza.SamzaException
import org.apache.samza.config._
import org.apache.samza.metrics.{MetricsRegistryMap, MetricsReporter, MetricsReporterFactory}
import org.apache.samza.serializers.{MetricsSnapshotSerdeV2, Serde, SerdeFactory}
import org.apache.samza.system.{SystemFactory, SystemProducer, SystemStream}
import org.apache.samza.util.ScalaJavaUtil.JavaOptionals
import org.apache.samza.util.{Logging, ReflectionUtil, StreamUtil, Util}

class MetricsSnapshotReporterFactory extends MetricsReporterFactory with Logging {

  protected def getProducer(reporterName: String, config: Config, registry: MetricsRegistryMap): SystemProducer = {
    val systemConfig = new SystemConfig(config)
    val systemName = getSystemStream(reporterName, config).getSystem
    val systemFactoryClassName = JavaOptionals.toRichOptional(systemConfig.getSystemFactory(systemName)).toOption
      .getOrElse(throw new SamzaException("Trying to fetch system factory for system %s, which isn't defined in config." format systemName))
    val systemFactory = ReflectionUtil.getObj(systemFactoryClassName, classOf[SystemFactory])

    info("Got system factory %s." format systemFactory)
    val producer = systemFactory.getProducer(systemName, config, registry)
    info("Got producer %s." format producer)

    producer
  }

  protected def getSystemStream(reporterName: String, config: Config): SystemStream = {
    val metricsConfig = new MetricsConfig(config)
    val metricsSystemStreamName = JavaOptionals.toRichOptional(metricsConfig.getMetricsSnapshotReporterStream(reporterName))
      .toOption
      .getOrElse(throw new SamzaException("No metrics stream defined in config."))
    val systemStream = StreamUtil.getSystemStreamFromNames(metricsSystemStreamName)
    info("Got system stream %s." format systemStream)
    systemStream
  }

  protected def getSerde(reporterName: String, config: Config): Serde[MetricsSnapshot] = {
    val streamConfig = new StreamConfig(config)
    val systemConfig = new SystemConfig(config)
    val systemStream = getSystemStream(reporterName, config)

    val streamSerdeName = streamConfig.getStreamMsgSerde(systemStream)
    val systemSerdeName = systemConfig.getSystemMsgSerde(systemStream.getSystem)
    val serdeName = streamSerdeName.orElse(systemSerdeName.orElse(null))
    val serializerConfig = new SerializerConfig(config)
    val serde = if (serdeName != null) {
      JavaOptionals.toRichOptional(serializerConfig.getSerdeFactoryClass(serdeName)).toOption match {
        case Some(serdeClassName) =>
          ReflectionUtil.getObj(serdeClassName, classOf[SerdeFactory[MetricsSnapshot]]).getSerde(serdeName, config)
        case _ => null
      }
    } else {
      new MetricsSnapshotSerdeV2
    }
    info("Got serde %s." format serde)
    serde
  }


  protected def getBlacklist(reporterName: String, config: Config): Option[String] = {
    val metricsConfig = new MetricsConfig(config)
    val blacklist = JavaOptionals.toRichOptional(metricsConfig.getMetricsSnapshotReporterBlacklist(reporterName)).toOption
    info("Got blacklist as: %s" format blacklist)
    blacklist
  }

  protected def getReportingInterval(reporterName: String, config: Config): Int = {
    val metricsConfig = new MetricsConfig(config)
    val reportingInterval = metricsConfig.getMetricsSnapshotReporterInterval(reporterName)
    info("Got reporting interval: %d" format reportingInterval)
    reportingInterval
  }

  protected def getJobId(config: Config): String = {
    val jobConfig = new JobConfig(config)
    jobConfig.getJobId
  }

  protected def getJobName(config: Config): String = {
    val jobConfig = new JobConfig(config)
    JavaOptionals.toRichOptional(jobConfig.getName).toOption
      .getOrElse(throw new SamzaException("Job name must be defined in config."))
  }


  def getMetricsReporter(reporterName: String, containerName: String, config: Config): MetricsReporter = {
    info("Creating new metrics snapshot reporter.")
    val registry = new MetricsRegistryMap

    val systemStream = getSystemStream(reporterName, config)
    val producer = getProducer(reporterName, config, registry)
    val reportingInterval = getReportingInterval(reporterName, config);
    val jobName = getJobName(config)
    val jobId = getJobId(config)
    val serde = getSerde(reporterName, config)
    val blacklist = getBlacklist(reporterName, config)

    val reporter = new MetricsSnapshotReporter(
      producer,
      systemStream,
      reportingInterval,
      jobName,
      jobId,
      containerName,
      Util.getTaskClassVersion(config),
      Util.getSamzaVersion,
      Util.getLocalHost.getHostName,
      serde, blacklist)

    reporter.register(this.getClass.getSimpleName, registry)

    reporter
  }
}

File: samza-core/src/main/java/org/apache/samza/metrics/reporter/MetricsSnapshotReporterFactory.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.samza.metrics.reporter;

import java.time.Duration;
import java.util.Optional;
import java.util.regex.Pattern;
import org.apache.samza.SamzaException;
import org.apache.samza.config.Config;
import org.apache.samza.config.JobConfig;
import org.apache.samza.config.MetricsConfig;
import org.apache.samza.config.SerializerConfig;
import org.apache.samza.config.StreamConfig;
import org.apache.samza.config.SystemConfig;
import org.apache.samza.metrics.MetricsRegistryMap;
import org.apache.samza.metrics.MetricsReporter;
import org.apache.samza.metrics.MetricsReporterFactory;
import org.apache.samza.serializers.MetricsSnapshotSerdeV2;
import org.apache.samza.serializers.Serde;
import org.apache.samza.serializers.SerdeFactory;
import org.apache.samza.system.SystemFactory;
import org.apache.samza.system.SystemProducer;
import org.apache.samza.system.SystemStream;
import org.apache.samza.util.ReflectionUtil;
import org.apache.samza.util.StreamUtil;
import org.apache.samza.util.SystemClock;
import org.apache.samza.util.Util;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


public class MetricsSnapshotReporterFactory implements MetricsReporterFactory {
  private static final Logger LOG = LoggerFactory.getLogger(MetricsSnapshotReporterFactory.class);

  @Override
  public MetricsReporter getMetricsReporter(String reporterName, String containerName, Config config) {
    LOG.info("Creating new metrics snapshot reporter.");
    MetricsRegistryMap registry = new MetricsRegistryMap();

    SystemStream systemStream = getSystemStream(reporterName, config);
    SystemProducer producer = getProducer(reporterName, config, registry);
    Duration reportingInterval = Duration.ofSeconds(getReportingInterval(reporterName, config));
    String jobName = getJobName(config);
    String jobId = getJobId(config);
    Serde<MetricsSnapshot> serde = getSerde(reporterName, config);
    Optional<Pattern> blacklist = getBlacklist(reporterName, config);

    MetricsSnapshotReporter reporter =
        new MetricsSnapshotReporter(producer, systemStream, reportingInterval, jobName, jobId, containerName,
            Util.getTaskClassVersion(config), Util.getSamzaVersion(), Util.getLocalHost().getHostName(), serde,
            blacklist, SystemClock.instance());
    reporter.register(this.getClass().getSimpleName(), registry);
    return reporter;
  }

  protected SystemProducer getProducer(String reporterName, Config config, MetricsRegistryMap registry) {
    SystemConfig systemConfig = new SystemConfig(config);
    String systemName = getSystemStream(reporterName, config).getSystem();
    String systemFactoryClassName = systemConfig.getSystemFactory(systemName)
        .orElseThrow(() -> new SamzaException(
            String.format("Trying to fetch system factory for system %s, which isn't defined in config.", systemName)));
    SystemFactory systemFactory = ReflectionUtil.getObj(systemFactoryClassName, SystemFactory.class);
    LOG.info("Got system factory {}.", systemFactory);
    SystemProducer producer = systemFactory.getProducer(systemName, config, registry);
    LOG.info("Got producer {}.", producer);
    return producer;
  }

  protected SystemStream getSystemStream(String reporterName, Config config) {
    MetricsConfig metricsConfig = new MetricsConfig(config);
    String metricsSystemStreamName = metricsConfig.getMetricsSnapshotReporterStream(reporterName)
        .orElseThrow(() -> new SamzaException("No metrics stream defined in config."));
    SystemStream systemStream = StreamUtil.getSystemStreamFromNames(metricsSystemStreamName);
    LOG.info("Got system stream {}.", systemStream);
    return systemStream;
  }

  protected Serde<MetricsSnapshot> getSerde(String reporterName, Config config) {
    StreamConfig streamConfig = new StreamConfig(config);
    SystemConfig systemConfig = new SystemConfig(config);
    SystemStream systemStream = getSystemStream(reporterName, config);

    Optional<String> streamSerdeName = streamConfig.getStreamMsgSerde(systemStream);
    Optional<String> systemSerdeName = systemConfig.getSystemMsgSerde(systemStream.getSystem());
    String serdeName = streamSerdeName.orElse(systemSerdeName.orElse(null));
    SerializerConfig serializerConfig = new SerializerConfig(config);
    Serde<MetricsSnapshot> serde;
    if (serdeName != null) {
      Optional<String> serdeFactoryClass = serializerConfig.getSerdeFactoryClass(serdeName);
      if (serdeFactoryClass.isPresent()) {
        SerdeFactory<MetricsSnapshot> serdeFactory = ReflectionUtil.getObj(serdeFactoryClass.get(), SerdeFactory.class);
        serde = serdeFactory.getSerde(serdeName, config);
      } else {
        serde = null;
      }
    } else {
      serde = new MetricsSnapshotSerdeV2();
    }
    LOG.info("Got serde {}.", serde);
    return serde;
  }

  protected Optional<Pattern> getBlacklist(String reporterName, Config config) {
    MetricsConfig metricsConfig = new MetricsConfig(config);
    Optional<String> blacklist = metricsConfig.getMetricsSnapshotReporterBlacklist(reporterName);
    LOG.info("Got blacklist as: {}", blacklist);
    return blacklist.map(Pattern::compile);
  }

  protected int getReportingInterval(String reporterName, Config config) {
    MetricsConfig metricsConfig = new MetricsConfig(config);
    int reportingInterval = metricsConfig.getMetricsSnapshotReporterInterval(reporterName);
    LOG.info("Got reporting interval: {}", reportingInterval);
    return reportingInterval;
  }

  protected String getJobId(Config config) {
    JobConfig jobConfig = new JobConfig(config);
    return jobConfig.getJobId();
  }

  protected String getJobName(Config config) {
    JobConfig jobConfig = new JobConfig(config);
    return jobConfig.getName().orElseThrow(() -> new SamzaException("Job name must be defined in config."));
  }
}
